src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs

```diff
@@ -76,14 +76,33 @@ internal sealed partial class SerializableOptionSet : OptionSet
         {
         }
 
+        /// <summary>
+        /// Returns an option set with all the serializable option values prefetched for given <paramref name="languages"/>,
+        /// while also retaining all the explicitly changed option values in this option set for any language.
+        /// </summary>
         public SerializableOptionSet WithLanguages(ImmutableHashSet<string> languages)
         {
             if (_languages.SetEquals(languages))
             {
                 return this;
             }
 
-            return _workspaceOptionSet.OptionService.GetSerializableOptionsSnapshot(languages);
+            // First create a base option set for the given languages.
```

Comment:
>// [](start = 12, length = 2)

Seems complicated. If we partitioned the _serializableOptionValues map per language, couldn't we just fetch the partitions of languages that were added from the option service? Instead of dealing each option separately.

Reply:
That is another implementation approach, and I think that has its own complexity. For example, for removal of language we want to make sure that we remove all the entries for that language _serializableOptionValues except the changed values. This felt much cleaner to me - start with all values that you would get if you created a fresh options set from global service and then just append changed option values.

Reply:
ok

---
In reply to: [392527102](https://github.com/dotnet/roslyn/pull/42388#discussion_r392527102) [](ancestors = 392527102)